app/services/account_search_service.rb

```diff
@@ -4,105 +4,126 @@ class AccountSearchService < BaseService
+      elsif domain_is_local?
+        Account.find_local(query_username)
+      else
+        Account.find_remote(query_username, query_domain)
+      end
+    end
+  end
 
-    return results if exact_match? && limit == 1
+  def search_results
+    return [] if limit_for_non_exact_results.zero?
 
-    results + search_results.to_a
+    @search_results ||= begin
+      if Chewy.enabled?
+        from_elasticsearch
+      else
+        from_database
+      end
+    end
   end
 
-  def query_blank_or_hashtag?
-    query.blank? || query.start_with?('#')
+  def from_database
+    if account
+      advanced_search_results
+    else
+      simple_search_results
+    end
   end
 
-  def split_query_string
-    @split_query_string ||= query.gsub(/\A@/, '').split('@')
+  def advanced_search_results
+    Account.advanced_search_for(terms_for_query, account, limit_for_non_exact_results, options[:following], offset)
   end
 
-  def query_username
-    @query_username ||= split_query_string.first || ''
+  def simple_search_results
+    Account.search_for(terms_for_query, limit_for_non_exact_results, offset)
   end
 
-  def query_domain
-    @query_domain ||= query_without_split? ? nil : split_query_string.last
-  end
+  def from_elasticsearch
+    must_clauses   = [{ multi_match: { query: terms_for_query, fields: %w(acct display_name), type: 'best_fields' } }]
+    should_clauses = []
 
-  def query_without_split?
-    split_query_string.size == 1
-  end
+    must_clauses   << { terms: { id: following_ids } } if account && options[:following]
+    should_clauses << { terms: { id: following_ids, boost: 2 } } if account
 
-  def domain_is_local?
-    @domain_is_local ||= TagManager.instance.local_domain?(query_domain)
-  end
+    query     = { bool: { must: must_clauses, should: should_clauses } }
+    functions = [reputation_score_function, followers_score_function, time_distance_function]
 
-  def search_from
-    options[:following] && account ? account.following : Account
-  end
+    records = AccountsIndex.query(function_score: { query: query, functions: functions, score_mode: 'avg' })
+                           .limit(limit_for_non_exact_results)
+                           .offset(offset)
+                           .objects
+                           .compact
 
-  def exact_match?
-    exact_match.present?
-  end
+    ActiveRecord::Associations::Preloader.new.preload(records, :account_stat)
 
-  def exact_match
-    return @exact_match if defined?(@exact_match)
+    records
+  end
 
-    @exact_match = begin
-      if domain_is_local?
-        search_from.without_suspended.find_local(query_username)
-      else
-        search_from.without_suspended.find_remote(query_username, query_domain)
-      end
-    end
+  def reputation_score_function
+    {
+      script_score: {
+        script: {
+          source: "doc['followers_count'].value / (doc['followers_count'].value + doc['following_count'].value + 1)",
```

Comment:
I am concerned about this, this favours people with more followers than people they follow, which, arguably, might avoid follow-bots, but will also favour celebrities-like and brands-like users. Also, that kind of stuff can be spoofed by remote attackers.

Reply:
It is also within a specific search query, where the user has an aim in mind, it is not a leaderboard. So the favouritism is in a very limited result space, and the impact is diminished by averaging with other scores. Spoofing would therefore target only specific searches.